Dubins planner

marcoCella opened the pull request.

marcoCella committed:
Dubins planner bt implementation

Still needs testing

marcoCella committed:
Sampling the dubins path only on the curves

Also removed not necessary waitings for services

marcoCella committed:
Sorting the dubins waypoints so that they follow the computed path

marcoCella committed:
Merge branch 'smarc-project:noetic-devel' into noetic-devel

marcoCella committed:
latlon to utm in degrees fix

marcoCella committed:
Working dubins

Still need to add the original waypoints to the path, for some reason they give problems

marcoCella committed:
Added original waypoints to dubins path

marcoCella committed:
Go away print

marcoCella committed:
Tuning

marcoCella committed:
Normalize quaternion

marcoCella committed:
Tuning

marcoCella committed:
Started working on inside turn

Not working for some reason

marcoCella committed:
Working inside curve

Just had to tweak the intersection tolerance

marcoCella committed:
Tuning

marcoCella committed:
Doc

marcoCella committed:
Autogen of the launch file

marcoCella committed:
Dynamic reconfigure turning radius and intersection radius

KKalem commented:
My good sir, there be a bug:

- Start everything as usual
- Run some waypoints, I tested with the new GUI
- After a couple WPs, probably when SAM is underwater already, send a mission
- dubins_mission_planner gets stuck with "Waiting for GPS Fix" message, and the entire system is unresponsive from then on.

When I check the services it is waiting for, they are there. If they were not present, the normal WP following wouldn't work either.

(also please remove trailing empty spaces, ty~)

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+
+        # Only continue if we have the gps fix
+        self.is_fix_ok = False
+        rospy.Subscriber("/lolo/core/gps", NavSatFix, self.gps_fix_cb)
```
Nice hardcoded sub ;)

Do you even need to check for gps fix _here_? I would argue to remove this check entirely, this is the job of the BT, with a higher-priority condition. So without a fix, a mission plan wont even be received.

marcoCella replied:
My bad, I hardcoded it for testing purposes and then I completely forgot to remove it 
If the bt takes care of checking the fix, I can definitely remove it

KKalem replied:
You can simply check if the vehicle object in the BB has a location in it. That is always the "last known pose". In theory, you don't even need to check that, because of that is not there, the BT wont even tick the mission parts, but being defensive never hurt anyone :)

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+
+        ll_to_utm_serv = self._get_latlon_to_utm_service()
+
+        # Wait for LoLo's first position before generating the path
```
also please avoid using robot-specific names in shared code, makes grepping harder~

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+        latlon_topic = "/" + robot_name + "/" + latlon_topic
+        gp = rospy.wait_for_message(latlon_topic, GeoPoint)
+        utm_x_lolo, utm_y_lolo = self.latlon_to_utm(gp.latitude, gp.longitude, gp.altitude, serv=ll_to_utm_serv, in_degrees=True)
+        lolo_utm = np.array([utm_x_lolo, utm_y_lolo])
```
i am assuming you are acquiring the current location of the vehicle to use as the initial point for dubins path, this is already available in the balckboard under the vehicle key (check bb_enums for exact name). Please use that instead.

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+        dubins_waypoints_final = dubins_waypoints_ordered[1:]
+
+        dubins_mission = MissionControl()
+        dubins_mission = mission
```
over-write?

marcoCella replied:
Personally I prefer having them as two separate objects, but I could definitely directly edit the original mission

KKalem replied:
but if you're gonna over-write the `dubins_mission` object with the `mission` one, why even construct the `MissionControl` object right before?

marcoCella replied:
Fair enough

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,212 @@ def path_to_list(self, path_msg):
+            #     dwp.travel_rpm = travel_rpm / 1.5
+            #     dwp.travel_speed = travel_speed / 1.5
+            
+            dwp.name = "dwp" + str(k)
```
Would be nice if you could somehow reference the two WPs this dwp is in between in its name, so that we have some idea where it is going without a graphical interface. IF it isn't too much of a hassle.

KKalem commented on smarc_bt/src/reconfig_server.py:
```diff
@@ -56,6 +56,16 @@ def __init__(self, config):
                                config.ENABLE_MANUAL_MISSION_LOG)
 
 
+        # dubins planner stuff
```
A checkbox to completely disable dubins-ing would be very useful for field debugging. Can be enabled by default. 
Also, turning radius is self-explanatory but intersection-radius isnt, maybe add a line before the default bit to explain it? (Adding `dubins_` prefix to both would also be useful to differentiate what they are for)

marcoCella replied:
https://github.com/smarc-project/smarc_missions/blob/b123784eb83f64cc18e50cc96c8cf53db9ca1261/smarc_bt/src/mission_plan.py#L262

There's a compute_dubins flag in mission_plan, I could make it dynamic reconfigurable so that it can be changed on the fly

KKalem replied:
dynamic is the way to go :+1:

marcoCella committed:
PR fixes

Removed trailing spaces, removed gps fix check, removed robot-specific names, current location now taken from blackboard, reference the original wps in the dwp name, dynamic reconfigure dubins flag

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,200 @@ def path_to_list(self, path_msg):
         return wps
 
 
+    def circle_line_segment_intersection(self, circle_center, circle_radius, pt1, pt2, full_line=False, tangent_tol=1e-5):
```
Both this and the dubins_mission_planner methods should really be in dubins.py, I'd rather keep the mission plan for wp-selection and wp-creation logic only. Similar to how coverage planner is its own thing.

KKalem replied:
Both this and the dubins_mission_planner methods should really be in dubins.py, I'd rather keep the mission plan for wp-selection and wp-creation logic only. Similar to how coverage planner is its own thing.

This also helps keep it contained and more free of side-effects. Especially given how large the functions are, its good to have a layer of separation to make it easier to keep in the head.

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,200 @@ def path_to_list(self, path_msg):
+            original_waypoints = waypoints_np.squeeze()  # Remove useless extra dimension
+
+        # Keep the other waypoints' parameters equal
+        mwp = mission.waypoints[0]
```
doesnt this basically set all dubins-wp's parameters to be the same as the _first_ user-given wp's params? Each wp in a mission can have unique params, this discards them all.

if a plan is A->B->C, and gets dubins'ed to be A - x,y,z -> B -p,q,r-> C x,y,z should have the params of B and p,q,r should have the params of C

KKalem commented on smarc_bt/src/mission_plan.py:
```diff
@@ -473,6 +539,200 @@ def path_to_list(self, path_msg):
+
+        ll_to_utm_serv = self._get_latlon_to_utm_service()
+
+        # Get the auv's current position to use as the intial point of the dubins path
```
this needs to happen when the mission is _started_, not _received_. You can move this logic to get_current_wp, there is already a check for "is this the first ever request of a wp?" in there.

KKalem reviewed (commented):
we'll get to individual corner cutting later.

marcoCella committed:
PR fix, moved stuff to dubins.py

marcoCella committed:
PR fix, wp details for each waypoint

marcoCella committed:
service name fix
